File: LeetCode/c++/RotateImage.cpp
#include <iostream>
#include <vector>

using namespace std;

void rotate(vector<vector<int>>& matrix) 
{
	int temp;
	int n = matrix.size();

	//先按水平线交换对折交换，接着通过对角线交换
	for (int i = 0; i < n; i++) //第i列
	{
		for (int j = 0; j < n/2; j++)
		{
			temp = matrix[j][i];
			matrix[j][i] = matrix[n-1-j][i];
			matrix[n - 1 - j][i] = temp;
		}
	}

	for (int i = 0; i < n; i++)
	{
		for (int j = 0; j < i; j++)
		{
			temp = matrix[i][j];
			matrix[i][j] = matrix[j][i];
			matrix[j][i] = temp;
		}
	}
}

vector<int> getImageRow(int n,int cnt)
{
	vector<int> v;

	for (int i = n; i < n + cnt; i++)
		v.push_back(i);

	return v;
}

void printV(vector<vector<int>> v)
{
	for (int i = 0; i < v.size(); i++)
	{
		for (int j = 0; j < v.size(); j++)
			cout << v[i][j] << " ";

		cout << endl;
	}
}

int main()
{
	vector<int> v1 = getImageRow(11, 5);
	vector<int> v2 = getImageRow(21, 5);
	vector<int> v3 = getImageRow(31, 5);
	vector<int> v4 = getImageRow(41, 5);
	vector<int> v5 = getImageRow(51, 5);

	vector<vector<int>> v;
	v.push_back(v1);
	v.push_back(v2);
	v.push_back(v3);
	v.push_back(v4);
	v.push_back(v5);

	printV(v);

	cout << endl;

	rotate(v);

	printV(v);

	return 0;
}

File: LeetCode/iter1/python/RotateImage.py
class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: void Do not return anything, modify matrix in-place instead.
        """
        #先按水平中线对折交换
        n = len(matrix)
        temp = 0
        half = n//2
        for i in range(0,n,1):#每列
            for j in range(0,half,1):
                temp = matrix[j][i]
                matrix[j][i] = matrix[n-1-j][i]
                matrix[n-1-j][i] = temp
                
        #斜对角交换
        for i in range(0,n,1):
            for j in range(0,i,1):
                temp = matrix[i][j]
                matrix[i][j] = matrix[j][i]
                matrix[j][i] = temp
   
                
def buildMatrix(start,cnt):
    matrix = []
    for i in range(start,start+cnt,1):
        matrix.append(i)
    
    return matrix
    
v1 = buildMatrix(11,5)
v2 = buildMatrix(21,5) 
v3 = buildMatrix(31,5) 
v4 = buildMatrix(41,5) 
v5 = buildMatrix(51,5)

v=[]
v.append(v1)
v.append(v2)
v.append(v3)
v.append(v4)
v.append(v5)

print(v)
       
s = Solution()
s.rotate(v)

print(v)